images.go

```diff
@@ -709,27 +710,58 @@ func (r *imageStore) Create(id string, names []string, layer, metadata string, c
 			return nil, fmt.Errorf("image name %q is already associated with image %q: %w", name, image.ID, ErrDuplicateName)
 		}
 	}
-	if created.IsZero() {
-		created = time.Now().UTC()
+	if options.CreationDate.IsZero() {
+		options.CreationDate = time.Now().UTC()
 	}
 
 	image = &Image{
 		ID:             id,
-		Digest:         searchableDigest,
-		Digests:        nil,
+		Digest:         options.Digest,
+		Digests:        copyDigestSlice(options.Digests),
 		Names:          names,
+		NamesHistory:   copyStringSlice(options.NamesHistory),
 		TopLayer:       layer,
-		Metadata:       metadata,
+		Metadata:       options.Metadata,
 		BigDataNames:   []string{},
 		BigDataSizes:   make(map[string]int64),
 		BigDataDigests: make(map[string]digest.Digest),
-		Created:        created,
-		Flags:          make(map[string]interface{}),
+		Created:        options.CreationDate,
+		Flags:          copyStringInterfaceMap(options.Flags),
 	}
 	err = image.recomputeDigests()
 	if err != nil {
 		return nil, fmt.Errorf("validating digests for new image: %w", err)
 	}
+	defer func() {
+		if err != nil {
+			// undo all that we might have done after this point
+			if e := os.RemoveAll(r.datadir(image.ID)); e != nil && !errors.Is(e, os.ErrNotExist) {
+				logrus.Debugf("while cleaning up data directory for image %q we failed to create: %v", id, e)
+			}
+			for _, digest := range image.Digests {
+				prunedList := imageSliceWithoutValue(r.bydigest[digest], image)
+				if len(prunedList) == 0 {
+					delete(r.bydigest, digest)
+				} else {
+					r.bydigest[digest] = prunedList
+				}
+			}
+			for _, name := range names {
+				delete(r.byname, name)
+			}
+			delete(r.byid, id)
+			r.idindex.Delete(id)
+			r.images = imageSliceWithoutValue(r.images, image)
+		}
+	}()
+	for _, item := range options.BigData {
```

Comment:
Absolutely non-blocking: A thing that _might_ make sense here, because unlike containers, images often have deterministic IDs, is to `os.RemoveAll(r.datadir(id))` here; that would ensure only the requested data is written.

OTOH it’s an extra syscall, and I’m not sure we ever write a `BigData` item that we wouldn’t want to write on a repeated attempt to create the same image, so I’m not all sure it would be useful.

If users are instructed to trigger `GarbageCollect` frequently enough, the `RemoveAll` wouldn’t buy us much.